### PR TITLE
Various small building improvements.

### PR DESCRIPTION
**Pull request by JacekHoleczek**
* Both building methods, the one using GNU make and the other one using CMake, should create ROOT dictionaries with the ".cxx" file extension (up to now, the first method was creating a ".cc" file). These automatically created source code files should not pollute the "src/" directory (following the general ROOT rules, they should be created in the current "build" directory).
* Copy all required files to the "external" build directory when the CMake building method is used.
* Protect `WCSim.cc` against undefined `WCSIMDIR`.

### CODE REVIEW

**Comment by pdeperio** [comment]
This is following [the discussion](https://github.com/nuPRISM/WCSim/pull/61#issuecomment-833124622) in https://github.com/nuPRISM/WCSim/pull/61

**Comment by JacekHoleczek** [comment]
Note: disregard the " GNUmakefile_* should make sure that WCSimRootDict.cxx is compiled as the first one" commit. It was a desperate trial to get reliable GNU make parallel builds, but it didn't really work, and I had to implement another fix (which seems to work fine).

**Comment by gpronost** [comment]
The commits look nice, however, I would like to also keep the deletion of the Dict files in src/ for some times, in order to automatically clean the directory of people updating their version to this one.

**Comment by JacekHoleczek** [comment]
Well, your big [bquilain/hybridPMT pull request to nuPRISM #61](https://github.com/nuPRISM/WCSim/pull/61) is not an ordinary set of patches. It adds files and changes the building procedure, so I think afterward, users should destroy their existing source and binaries trees and start from scratch again.

**Comment by JacekHoleczek** [comment]
Well, let's hope the GNU make parallel builds are finally safe (without any dirty tricks).

**Comment by gpronost** [comment]
> Well, your big [bquilain/hybridPMT pull request to nuPRISM #61](https://github.com/nuPRISM/WCSim/pull/61) is not an ordinary set of patches. It adds files and changes the building procedure, so I think afterward, users should destroy their existing source and binaries trees and start from scratch again.

Well, asking to delete the sources instead of pulling is not really how github should be used. Anyway, I was mostly thinking of bquilain branch users not of nuPRISM branch users.

**Comment by JacekHoleczek** [comment]
Well, I had this issue many times with ROOT. I learned that the safest way is to destroy everything and start from scratch again.
However, if someone wants to do it the "canonical way" ... one should first execute "`make clean`" (with the old existing makefiles, so that any existing derived files disappear, e.g., "`src/*Dict*`"), then pull the new source code, and finally configure and make the new sources (with new makefiles).

**Comment by pdeperio** [comment]
Well, can you all comment directly in the code diff? It's hard to follow what exactly you're talking about ;)

**Comment by bquilain** [comment]
Hi @JacekHoleczek and @gpronost . Let me say same as @pdeperio : could you please point out which part of the code you are talking about please? It would be really helpful to understand :)

File: make_sukap_ROOT5.sh [hunk 1]
```diff
@@ -16,7 +16,7 @@ if [ ! -d ${build_directory} ]; then
 	if [ -d ${G4WORKDIR} ]; then
 		rm -r ${G4WORKDIR}
 	fi
-	rm *.o *.a *.so *~ */*~ src/*Dict*
```
Comment (gpronost) [inline make_sukap_ROOT5.sh:19]: Here

File: make_sukap_ROOT6.sh [hunk 2]
```diff
@@ -16,7 +16,7 @@ if [ ! -d ${build_directory} ]; then
 	if [ -d ${G4WORKDIR} ]; then
 		rm -r ${G4WORKDIR}
 	fi
-	rm *.o *.a *.so *~ */*~ src/*Dict*
```
Comment (gpronost) [inline make_sukap_ROOT6.sh:19]: Here

File: GNUmakefile_full [hunk 3]
```diff
@@ -100,12 +97,13 @@ clean_wcsim:
 #Clean with :: will be executed before the normal G4 clean
 #RootDict may lead to some troubles if not clean before checking dependencies
 clean::
-	@echo "Cleaning RootDict ..."
-	@rm -f ./WCSimRootDict.*
-	@rm -f ./src/WCSimRootDict.h ./src/WCSimRootDict.cc ./src/WCSimRootDict_rdict.pcm WCSimRootDict_rdict.pcm
```
Comment (gpronost) [inline GNUmakefile_full:105]: Here

**Comment by JacekHoleczek** [comment]
It's about the "clean" target in "GNUmakefile_*" (I try to say that it should NOT know old / redundant files as the user is expected to execute "`make clean`" before updating the source code)

**Comment by bquilain** [comment]
Thank you very much @JacekHoleczek and @gpronost ! It seems good to me, but then, what is the motivation for not keeping the src/*Dict* cleaner? I understand what users are suposed to do (make clean), but we know well that users act differently that they should (us included). Is there any good motivation for not keeping this cleaning, apart from simplicity?

**Comment by JacekHoleczek** [comment]
The "clean" target should only remove files that were really created by "make".
It must not delete other files, which a user may have created (added) on purpose.

**Comment by gpronost** [comment]
Well, if one user decide to create a WCSimRootDict files in src, it will likely not compile properly. 
But anyway, as you want, I will make a clean up script to solve potential troubles for users forgetting to clean their directory.

**Comment by JacekHoleczek** [comment]
Actually, also "`git status`" is helpful here. It will show any modified and / or added (local) files, so a user may easily find anything that does not belong into the source code tree (of course, it requires that the "`.gitignore`" file is properly maintained, too).

**Comment by gpronost** [comment]
> Actually, also "`git status`" is helpful here. It will show any modified and / or added (local) files, so a user may easily see anything that does not belong to it.

In this particular case, `git status` will not indicate anything, since we have this in `.gitignore`:
`*WCSimRootDict*
`

**Comment by JacekHoleczek** [comment]
Oops. So, "`make clean`" is really mandatory (unless we try to implement a better pattern in the "`.gitignore`").

**Comment by JacekHoleczek** [comment]
Well, there exist two shell scripts "`make_sukap_ROOT5.sh`" and "`make_sukap_ROOT6.sh`". They do not use "`make clean`" but delete files listed in place. So, maybe one could add these old / redundant files there.
Do you think this would be sufficient?

**Comment by gpronost** [comment]
Actually, I am planning to merge these scripts and to move them, I can make a clean up script (which could be called by these scripts) to take care of the old files.

**Comment by JacekHoleczek** [comment]
So, I fixed the "`.gitignore`" patterns.

File: make_sukap_ROOT5.sh (hunk 1 above)
Comment (pdeperio) [reply make_sukap_ROOT5.sh:19]: i guess i couldn't follow the discussion. looks like `*Dict*` was just moved to the root directory, but is still cleaned?

Comment (gpronost) [reply make_sukap_ROOT5.sh:19]: WCSimRootDict was moved from `./src/` to `./`, but for some users, WCSimRootDict is currently still in their `./src/`. The modification here, and in the other scripts and GNUmakefile, removes the deletion of the WCSimRootDict in `./src/`. So, if a user still have  WCSimRootDict in `./src/` and updates WCSim without cleaning it first, it will failed to compile, and will need to manually remove WCSimRootDict in  `./src/`. 

I wanted to keep this removal for some times, in order to give time to user to clean their directories automatically after the update. But @JacekHoleczek is against this, and points that we should not remove files which are not created during the compilation (I agree with him on this point, but I think cleaning old files is an acceptable exception to this rule).

Anyway, let's go with the current version, and I will make another PR to merge the compilation scripts for cmake and add a script to clean up old files like src/WCSimRootDict*

**Comment by bquilain** [comment]
> So, I fixed the "`.gitignore`" patterns.

Thank you very much

**Comment by pdeperio** [comment]
@bquilain @gpronost is this ok to merge now?

**Comment by bquilain** [comment]
@pdeperio Yes I think, I am happy with this. @gpronost do you have any worry if we merge now?

**Comment by gpronost** [comment]
It's fine for me, I will make some scripts to cleanup after :)
